### PR TITLE
Automated version bumping and validation

### PR DESCRIPTION
**Pull request by tamerbader**
### Version Bumping Script
- Created a new `update-sdk-version.rb` script that reads from `sdk-constants.rb`(prev `podspec_constants.rb`) to update `SquareMobilePaymentsSDK.podspec`, `MockReaderUI.podspec`, `Package.swift`, and `README` files with the new values.
- Created a sdk-utlis.rb file that contains helper classes for other scripts in the repo.
- Gemfile is needed to ensure all gems required by the scripts are installed during `bundle install`.
- `.ruby-version` is needed to lock ruby version. 
### Github PR Action Validation
- This PR also defines a Github PR Action in `.github/workflows/validate-pull-request.yml` to validate PR changes being merged into main. 
- The github action job runs the script found in `Scripts/git/pull-request-validator.rb` to perform validations.
- The script relies on running `pod spec lint` to validate all podspecs in the repo. This command ensures the podspec is properly formatted and that the frameworks are able to be found.
- The script also runs `swift package resolve` which allows us to validate the SPM specs in Package.swift. This command will attempt to resolve the packages ensuring the xcframeworks for all the packages are able to be found by SPM.
- The script finally builds all the templated files using the defined constants and stores the outputs into a tmp directory. It diffs the tmp files with the ones being checked in to ensure there is no difference. If a difference is noted, then that indicates a developer made changes to the actual output file instead of the template file.
- If any one of these validations fail, then the github action fails preventing the user from merging their PR. They will need to look at the logs of the job  for more details on how to fix.

### Script Instructions
1. Run `bundle install` to download all gems
2. Make changes to version number and commit sha in `sdk-constants.rb`
3. Run `ruby Scripts/update-sdk-version.rb` to update all files with correct data.
4. Check in all file changes
5. Run `ruby Scripts/git/pull-request-validator.rb` locally. This will run the github action script locally and ensure your changes are valid before pushing.
6. Push your changes up to the branch and create a PR against main
7. In the newly opened PR, you should see a github action workflow running to validate the PR.

### Demo

Test Version Update Script

https://github.com/square/mobile-payments-sdk-ios/assets/20408179/ad43479c-fc25-4384-9757-7831bdeba912

[Test Github Action PR Success
](https://github.com/square/mobile-payments-sdk-ios/actions/runs/9766016883/job/26958024303?pr=7)

[Test Github Action PR Fail - Bad Commit SHA](https://github.com/square/mobile-payments-sdk-ios/actions/runs/9766809074/job/26960597013?pr=9)

[Test Github Action PR Fail - Modifying Output File Not Template](https://github.com/square/mobile-payments-sdk-ios/actions/runs/9766459914/job/26959460268?pr=9)

### CODE REVIEW

File: Scripts/git/pre-commit-validator.rb [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+# Generate Temporary Package.swift
+template_builder.build_and_write('./Scripts/templates/Package.swift.erb', 'Package.swift', './tmp/')
+
+# Validates no changes in the diff between generated files and the files being checked in
+if FileComparator.compare('./README.md', './tmp/README.md') &&
+	 FileComparator.compare('./SquareMobilePaymentsSDK.podspec', './tmp/SquareMobilePaymentsSDK.podspec') &&
+	 FileComparator.compare('./MockReaderUI.podspec', './tmp/MockReaderUI.podspec') &&
+	 FileComparator.compare('./Package.swift', './tmp/Package.swift')
+	# All files are identical!
+	delete_temp_dir
```
Comment (plinio-square) [inline Scripts/git/pre-commit-validator.rb:76]: you're always calling `delete_temp_dir` so might as well move it out of the `if` statement

File: Scripts/git/pre-commit-validator.rb [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+class FileComparator
+  def self.compare(file1_path, file2_path)
+    unless File.exist?(file1_path) && File.exist?(file2_path)
+      return false
+    end
+
+    file1 = File.open(file1_path)
+    file2 = File.open(file2_path)
+
+    file1.each_line.zip(file2.each_line) do |line1, line2|
```
Comment (plinio-square) [inline Scripts/git/pre-commit-validator.rb:14]: you can use `FileUtils.identical?(file1, file2)` instead

Comment (plinio-square) [reply Scripts/git/pre-commit-validator.rb:14]: I wouldn't completely remove this method though, you might want to be able to throw an error if the file isn't found but once you have a file handler you can just use the mentioned method

File: Scripts/sdk_utilities.rb [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    renderer = ERB.new(template)
+    result = renderer.result(get_binding)
+    File.open("#{output_dir}#{file_name}", 'w+') do |file|
+      file.write(result)
+    end
+  end
+
+  private
+
+  def get_binding
```
Comment (plinio-square) [inline Scripts/sdk_utilities.rb:25]: do you need this? I think `binding` should be available in this context already, the common case for a getter for binding is to make it public (which I don't think you'd need?)

**Comment by plinio-square** [comment]
lgtm just had some small suggestions

**Comment by kirstiebooras** [comment]
Couple of questions.
1. I modified the sdk_constants.rb to have a commit sha of "bad_commit". I then ran: `ruby Scripts/update-sdk-version.rb && git add . && git commit -m "bad commit should fail"` I got the message `❎ Pre-Commit Validation Failed. Aborting Commit` but the commit did in fact succeed and wasn't aborted. `git status` shows `Your branch is ahead of 'origin/tbader/automating-release-script' by 1 commit` Is that expected? (Note: I did the same thing with a bad branch and the commit succeeded as well)
2. The error message I got was 

>  -> MockReaderUI (2.0.0-beta1)
    - ERROR | [iOS] unknown: Encountered an unknown error (/bin/bash -c 
set -e
unzip XCFrameworks/MockReaderUI_bad_commit.zip
unzip:  cannot find or open XCFrameworks/MockReaderUI_bad_commit.zip, XCFrameworks/MockReaderUI_bad_commit.zip.zip or XCFrameworks/MockReaderUI_bad_commit.zip.ZIP.
) during validation.

Is there any way we can make that more explicit? i.e. `Encountered an error: Unknown commit "bad_commit"`? Or is that fairly difficult to do?

File: Scripts/git/pre-commit-validator.rb [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+template_builder = TemplateBuilder.new(
+	SquareMobilePaymentsSDK::VERSION,
+	SquareMobilePaymentsSDK::COMMIT_SHA,
+	SquareMobilePaymentsSDK::LICENSE,
+	SquareMobilePaymentsSDK::HOMEPAGE_URL,
+	SquareMobilePaymentsSDK::AUTHORS,
+	SquareMobilePaymentsSDK::IOS_DEPLOYMENT_TARGET
+)
+
+# Generate Temporary README
```
Comment (tobischw) [inline Scripts/git/pre-commit-validator.rb:58]: I feel like line 58-68, the .erb files could be kept track of in an array constants somewhere, and the target location could be inferred automatically based off the path name (except no `.erb` at the end)

File: Scripts/git/pre-commit-validator.rb (hunk 2 above)
Comment (tamerbader) [reply Scripts/git/pre-commit-validator.rb]: Updated 🙏

File: Scripts/sdk_utilities.rb (hunk 3 above)
Comment (tamerbader) [reply Scripts/sdk_utilities.rb]: Removed! 🙏

**Review by tobischw — APPROVED** [review]
great work on this!

**Review by plinio-square — APPROVED** [review]

**Comment by plinio-square** [comment]
good job, Tamer

**Review by kirstiebooras — APPROVED** [review]
Yay! I'm not getting that issue anymore.

File: .github/workflows/validate-pull-request.yml [hunk 5]
```diff
@@ -0,0 +1,27 @@
+name: MPSDK Pull Request Validator
+
+on:
+  pull_request:
```
Comment (brandonjenniges) [inline .github/workflows/validate-pull-request.yml:4]: I wonder if we should also add it to check on a push to main also? Just to make sure nothing broke post merge?

```
push:
    branches:
      - main
```

**Review by brandonjenniges — APPROVED** [review]
LGTM!

File: .github/workflows/validate-pull-request.yml (hunk 5 above)
Comment (tamerbader) [reply .github/workflows/validate-pull-request.yml:4]: Yeah i'm down to add that as well!

**Pull request merged**